libraries/standard/mqtt/src/iot_mqtt_serialize.c

```diff
@@ -1483,454 +2019,575 @@ IotMqttError_t _IotMqtt_DeserializePuback( _mqttPacket_t * pPuback )
+    IOT_FUNCTION_EXIT_NO_CLEANUP();
 }
 
 /*-----------------------------------------------------------*/
 
-IotMqttError_t _IotMqtt_DeserializePingresp( _mqttPacket_t * pPingresp )
+IotMqttError_t IotMqtt_SerializePingreq( uint8_t * pBuffer,
+                                         size_t bufferSize )
 {
     IOT_FUNCTION_ENTRY( IotMqttError_t, IOT_MQTT_SUCCESS );
+    uint8_t * pPingreqPacket = NULL;
+    size_t packetSize = 0;
 
-    /* Check that the control packet type is 0xd0. */
-    if( pPingresp->type != MQTT_PACKET_TYPE_PINGRESP )
+    if( pBuffer == NULL )
     {
-        IotLog( IOT_LOG_ERROR,
-                &_logHideAll,
-                "Bad control packet type 0x%02x.",
-                pPingresp->type );
-
-        IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_RESPONSE );
+        IotLogError( "IotMqtt_SerializePingreq() called with NULL buffer pointer." );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_PARAMETER );
     }
-    else
+
+    if( bufferSize < MQTT_PACKET_PINGREQ_SIZE )
     {
-        EMPTY_ELSE_MARKER;
+        IotLogError( "Pingreq length (%lu) exceeds buffer size (%lu).",
+                     MQTT_PACKET_DISCONNECT_SIZE, bufferSize );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_PARAMETER );
     }
 
-    /* Check the "Remaining length" (second byte) of the received PINGRESP. */
-    if( pPingresp->remainingLength != MQTT_PACKET_PINGRESP_REMAINING_LENGTH )
+    /* Call internal function with local varaibles, as ping request uses
+     * static memory, there is no need to pass the buffer
+     * Note: _IotMqtt_SerializePingReq always succeeds */
+    _IotMqtt_SerializePingreq( &pPingreqPacket, &packetSize );
+    memcpy( pBuffer, pPingreqPacket, MQTT_PACKET_PINGREQ_SIZE );
+    IOT_FUNCTION_EXIT_NO_CLEANUP();
+}
+
+/*-----------------------------------------------------------*/
+
+IotMqttError_t IotMqtt_DeserializePublish( IotMqttPacketInfo_t * pMqttPacket )
+{
+    IOT_FUNCTION_ENTRY( IotMqttError_t, IOT_MQTT_SUCCESS );
+    /* Internal MQTT packet structure */
+    _mqttPacket_t mqttPacket;
+    /* Internal MQTT operation structure needed for deserializing publish */
+    _mqttOperation_t mqttOperation;
+
+    if( pMqttPacket == NULL )
     {
-        IotLog( IOT_LOG_ERROR,
-                &_logHideAll,
-                "PINGRESP does not have remaining length of %d.",
-                MQTT_PACKET_PINGRESP_REMAINING_LENGTH );
+        IotLogError( "IotMqtt_DeserializePublish()called with NULL pMqttPacket pointer." );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_PARAMETER );
+    }
 
-        IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_RESPONSE );
+    if( ( pMqttPacket->type & 0xf0 ) != MQTT_PACKET_TYPE_PUBLISH )
+    {
+        IotLogError( "IotMqtt_DeserializePublish() called with incorrect packet type:(%lu).", pMqttPacket->type );
+        IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_PARAMETER );
     }
-    else
+
+    /* Set intenal mqtt packet parameters. */
```

Comment:
spelling: _internal_, maybe search throughout for the misspelled version since this is the second of these I've found.